File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode * temp=head;
        int cnt=0;
        while(temp)
        {
            temp=temp->next;
            cnt++;
        }
        temp=head;
        int tar=0;
        while(temp)
        {
            if(tar<cnt/2)
            {
                tar++;
                temp=temp->next;
            }
            else{
                break;
            }
            
        }
        ListNode * x=temp;
       return x;

    }
};

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        // ❗️Correct condition: check fast != null first to avoid NullPointerException
        while (fast != null && fast.next != null) {
            slow = slow.next;         // move 1 step
            fast = fast.next.next;    // move 2 steps
        }

        return slow;
    }
}
